### PR TITLE
[Sei] refactor order representation by order ID

### PR DESCRIPTION
**Pull request by codchen**
PR is large because references to orders are ubiquitous in `dex` module. There are no changes to the order matching core logic though, and most changes are just refactor of data structures.

- Store individual orders by order ID (uint64). Specifically, all orders and cancels will be stored immutably, and each account has an index to a list of active order IDs. New orders will have IDs added to the index, whereas full executions and cancellation will have IDs removed from the index.
- Unified order representation as `order.pb.go` and deprecated `OrderPlacement`, `dexcache.Order`, `LiquidationOrder`, and `ContractOrderPlacement`.
- Store order IDs instead of regex strings as allocation identifiers on the order book.
- Hide product-specific details like `leverage` and `position effect` in an opaque `Data` field that will be passed through to the contract. In other words, `dex` doesn't need to know about `leverage` etc. anymore.

A corresponding change will be made on Vortex's side and verified via integration tests. This PR should not be merged until that happens. [EDIT: verified in https://github.com/sei-protocol/vortex-contract/pull/78]

### CODE REVIEW

File: proto/dex/order.proto
Comment (LCyson) [inline proto/dex/order.proto:22]: how does other projects interact with order placement rn

File: x/dex/keeper/msg_server_place_orders.go
Comment (LCyson) [inline x/dex/keeper/msg_server_place_orders.go:51]: is this a uuid?

File: x/dex/keeper/order.go
Comment (LCyson) [inline x/dex/keeper/order.go:1]: can we add a test for the keeper store

File: proto/dex/order_entry.proto
Comment (LCyson) [inline proto/dex/order_entry.proto:28]: ?

File: x/dex/module.go
Comment (LCyson) [inline x/dex/module.go:292]: what's zeroorders here

File: x/dex/keeper/msg_server_place_orders.go
Comment (LCyson) [reply x/dex/keeper/msg_server_place_orders.go:51]: ever-increasing uint64

**Comment by LCyson** [comment]
TLDR, options for upgrading:
1. In migration handler, clear all previous states [Best option]
2. Upgrade before anyone starts using the dex module on incentivized testnet
3. Complete migration handler logic
4. Start a new chain

File: x/dex/module.go
Comment (codchen) [reply x/dex/module.go:292]: added a comment

File: x/dex/keeper/order.go
Comment (codchen) [reply x/dex/keeper/order.go:1]: added

File: proto/dex/order.proto
Comment (LCyson) [inline proto/dex/order.proto:50]: what's the data here

File: utils/slice.go
Comment (LCyson) [inline utils/slice.go:1]: can we add a test for this

File: x/dex/types/keys.go
Comment (LCyson) [inline x/dex/types/keys.go]: nit: why not merge the these two const closure as both of they hold constants for key

File: x/dex/cache/cache.go
Comment (LCyson) [inline x/dex/cache/cache.go:1]: can we add a test for this file

File: x/dex/cache/cache.go
Comment (LCyson) [inline x/dex/cache/cache.go]: can we move this to the top of the `MemState` definition

File: x/dex/module.go
Comment (LCyson) [inline x/dex/module.go:271]: random q, theoretically this memstate is still redundant as the keeper also doesn't store directly to disk right? what's the reason we're keeping the MemState layer rn?

File: x/dex/types/alias.go
Comment (LCyson) [inline x/dex/types/alias.go:1]: can we add a test for this file

File: x/dex/types/keys.go
Comment (LCyson) [inline x/dex/types/keys.go:79]: nit: add tests

**Review by LCyson — CHANGES_REQUESTED** [review]
lgtm pending small tests

File: proto/dex/order.proto
Comment (codchen) [reply proto/dex/order.proto:50]: any product-specific data that a contract needs but dex doesn't need to know

File: x/dex/module.go
Comment (codchen) [reply x/dex/module.go:271]: we need memstate just to retain information from `DeliverTx` to `EndBlock`. It can be discarded after `EndBlock` runs

File: x/dex/types/alias.go
Comment (codchen) [reply x/dex/types/alias.go:1]: added

File: x/dex/types/keys.go
Comment (codchen) [reply x/dex/types/keys.go:79]: added

File: x/dex/cache/cache.go
Comment (codchen) [reply x/dex/cache/cache.go]: :+1:

File: x/dex/cache/cache.go
Comment (codchen) [reply x/dex/cache/cache.go:1]: added

File: x/dex/types/keys.go
Comment (codchen) [reply x/dex/types/keys.go]: :+1:

File: utils/slice.go
Comment (codchen) [reply utils/slice.go:1]: added

File: proto/dex/order.proto
Comment (LCyson) [reply proto/dex/order.proto:50]: why this is a string in that case, shouldn't this be an object with needed fields? or the idea here is to parse the data string into a json, does it guarantee backwards/forwards compatibility in cosmwasm?

**Review by LCyson — APPROVED** [review]

File: proto/dex/order.proto
Comment (LCyson) [reply proto/dex/order.proto:50]: oh saw it in the other pr, will ask offline if i still have q after that

**Pull request merged**
